Source/Base/ASBaseDefines.h

```diff
@@ -245,26 +229,38 @@
  * Create a new set by mapping `collection` over `work`, ignoring nil.
  */
 #define ASSetByFlatMapping(collection, decl, work) ({ \
-  NSMutableSet *s = [NSMutableSet set]; \
+  CFTypeRef _cArray[collection.count]; \
+  NSUInteger _i = 0; \
   for (decl in collection) {\
-    id result = work; \
-    if (result != nil) { \
-      [s addObject:result]; \
+    if ((_cArray[_i] = (__bridge_retained CFTypeRef)work)) { \
+      _i++; \
     } \
   } \
-  s; \
+  NSSet *result; \
+  if (_i == 0) { \
+    /** Zero fast path. */ \
+    result = [NSSet set]; \
+  } else if (_i == 1) { \
+    /** NSSingleObjectSet is fast. Create one and release. */ \
+    CFTypeRef val = _cArray[0]; \
+    result = [NSSet setWithObject:(__bridge id)val]; \
+    CFBridgingRelease(val); \
+  } else { \
+    CFSetCallBacks cb = kCFTypeSetCallBacks; \
+    cb.retain = NULL; \
+    result = (__bridge NSSet *)CFSetCreate(kCFAllocatorDefault, _cArray, _i, &cb); \
+  } \
+  result; \
 })
 
 /**
  * Create a new ObjectPointerPersonality NSHashTable by mapping `collection` over `work`, ignoring nil.
  */
 #define ASPointerTableByFlatMapping(collection, decl, work) ({ \
```

Comment:
Is this one not yet used? Just wanted to say, I think this #define-based approach is very powerful as an optimization, in its ability to avoid wrapping and retains that most other mechanisms lead to. We should use it judiciously, but I bet there are a few more cases where it is justified and impactful.